.github/workflows/check-trailing-whitespace.yml

```diff
@@ -39,8 +39,13 @@ jobs:
 
           # Check for missing EOF newline
           if [ -f "$file" ] && [ -s "$file" ]; then
-            last_char=$(tail -c 1 "$file")
-            if [ "$last_char" != "" ] && [ "$last_char" != $'\n' ]; then
+            # I know it looks really dumb adding an x to the output here but it's required because
+            # command substitution will strip ALL trailing newlines, but we're exactly interested in trailing newlines!
+            # We can trick it by ensuring that the output of the command substitution does not end in a newline, thus
+            # preserving any newlines before the 'x'.
+            # See https://bash-hackers.gabe565.com/syntax/expansion/cmdsubst/
+            last_char=$(tail -c 1 "$file"; echo -n "x")
+            if [ "$last_char" != "x" ] && [ "$last_char" != $'\nx' ]; then
```

Comment:
you can write this test as `[[ -n "$(tail -c 1 <"$file")" ]]` instead of doing this x at the end thing

Reply:
I wasn't sure if the shell in the action was running bash or not. I'll give it a try.

Reply:
This specifically relies on the fact that the command substitution will strip the newline... which I agree works but it feels bad and it not very clear what is happening to anyone who looks at it.

I personally would like a solution that explicitly looks at the last character and says "yes this is a newline character"